src/OpenTelemetry.AutoInstrumentation.StartupHook/RulesEngine/NativeProfilerDiagnosticsRule.cs

```diff
@@ -0,0 +1,112 @@
+// </copyright>
+
+using OpenTelemetry.AutoInstrumentation.Helpers;
+using OpenTelemetry.AutoInstrumentation.Logging;
+
+namespace OpenTelemetry.AutoInstrumentation.RulesEngine;
+
+/// <summary>
+/// Diagnoses native profiler issues.
+/// Diagnoser only supports .NET (Core) runtime.
+/// </summary>
+internal class NativeProfilerDiagnosticsRule : Rule
+{
+    private const string ProfilerEnabledVariable = "CORECLR_ENABLE_PROFILING";
+    private const string ProfilerIdVariable = "CORECLR_PROFILER";
+    private const string ProfilerPathVariable = "CORECLR_PROFILER_PATH";
+    private const string Profiler32BitPathVariable = "CORECLR_PROFILER_PATH_32";
+    private const string Profiler64BitPathVariable = "CORECLR_PROFILER_PATH_64";
+    private const string ProfilerId = "{918728DD-259F-4A6A-AC2B-B85E1B658318}";
+
+    private static readonly IOtelLogger Logger = OtelLogging.GetLogger("StartupHook");
+
+    public NativeProfilerDiagnosticsRule()
+    {
+        Name = "Native profiler diagnoser";
+        Description = "Verifies that native profiler is correctly setup in case it's enabled.";
+    }
+
+    internal override bool Evaluate()
+    {
+        var isProfilerEnabled = EnvironmentHelper.GetEnvironmentVariable(ProfilerEnabledVariable) == "1";
+        if (!isProfilerEnabled)
+        {
+            Logger.Debug($"CLR profiler is not enabled.");
+            return true;
+        }
+
+        var profilerId = EnvironmentHelper.GetEnvironmentVariable(ProfilerIdVariable);
+        if (profilerId != ProfilerId)
+        {
+            Logger.Warning($"Detected CLR profiler is enabled but different profiler ID is provided '{profilerId}'.");
+
+            // Different native profiler not assosiated to OTel might be used. We don't want to fail here.
+            return true;
+        }
+
+        var profilerPath = EnvironmentHelper.GetEnvironmentVariable(ProfilerPathVariable);
+        var profilerPath32 = EnvironmentHelper.GetEnvironmentVariable(Profiler32BitPathVariable);
+        var profilerPath64 = EnvironmentHelper.GetEnvironmentVariable(Profiler64BitPathVariable);
+
+        var isPathUndefined = string.IsNullOrWhiteSpace(profilerPath);
+        var isPath32Undefined = string.IsNullOrWhiteSpace(profilerPath32);
+        var isPath64Undefined = string.IsNullOrWhiteSpace(profilerPath64);
+
+        if (isPathUndefined)
+        {
+            if (isPath32Undefined && isPath64Undefined)
+            {
+                Logger.Error($"CLR profiler path is not defined. Define '{ProfilerPathVariable}', '{Profiler32BitPathVariable}' or '{Profiler64BitPathVariable}'.");
+                return false;
+            }
+
+            if (Environment.Is64BitProcess && isPath64Undefined)
+            {
+                Logger.Error($"CLR profiler (64bit) path is not defined. Define '{Profiler64BitPathVariable}'.");
+                return false;
+            }
+
+            if (!Environment.Is64BitProcess && isPath32Undefined)
```

Comment:
Comment here for clarity: this is a check for 32-bit. Negations in the code may be easily overlooked during review.

Reply:
Refactored.